src/System.Management.Automation/engine/CommandCompletion/CompletionAnalysis.cs

```diff
@@ -1025,6 +1065,221 @@ private static string GetFirstLineSubString(string stringToComplete, out bool ha
+            {
+                if (type.IsEnum)
+                {
+                    completions = GetResultForEnum(type, completionContext);
+                    return true;
+                }
+            }
+
+            // If the assignment itself was unconstrained, the variable still might be
+            if (!TryGetTypeConstraintOnVariable(completionContext, variableAst.VariablePath.UserPath, out typeConstraint, out setConstraint))
+            {
+                return false;
+            }
+
+            // Again try the [ValidateSet()] constraint first
+            if (setConstraint?.ValidValues != null)
+            {
+                completions = GetResultForSet(type, setConstraint.ValidValues, completionContext);
+                return true;
+            }
+
+            // Then try to complete for an enum type
+            if (!typeConstraint.IsEnum)
+            {
+                return false;
+            }
+            else
+            {
+                completions = GetResultForEnum(typeConstraint, completionContext);
+                return true;
+            }
+        }
+
+        private static List<CompletionResult> GetResultForSet(
+            Type type,
+            IList<string> validValues,
+            CompletionContext completionContext)
+        {
+            var allValues = new List<string>();
+            foreach (string value in validValues)
+            {
+                if (type == typeof(string) || type.IsEnum)
+                {
+                    allValues.Add(GetQuotedString(value, completionContext));
+                }
+                else
+                {
+                    allValues.Add(value);
+                }
+            }
+
+            return GetMatchedResults(allValues, completionContext);
+        }
+
+        private static List<CompletionResult> GetMatchedResults(
+            List<string> allValues,
+            CompletionContext completionContext)
+        {
+            var stringToComplete = string.Empty;
+            if (completionContext.TokenAtCursor != null)
+            {
+                stringToComplete = completionContext.TokenAtCursor.Text;
+            }
+
+            string matchString = stringToComplete + "*";
+            var wildcardPattern = WildcardPattern.Get(matchString, WildcardOptions.IgnoreCase | WildcardOptions.CultureInvariant);
+
+            IEnumerable<string> matchedResults = allValues.Where(r => wildcardPattern.IsMatch(r));
+            if (matchedResults == null || !matchedResults.Any())
+            {
+                matchedResults = allValues;
+            }
+
+            var result = new List<CompletionResult>();
+            foreach (var match in matchedResults)
+            {
+                result.Add(new CompletionResult(match));
+            }
+
+            return result;
+        }
+
+        private static string GetQuotedString(
+            string value,
+            CompletionContext completionContext)
+        {
```

Comment:
Is it possible to reuse `CompletionCompleters.HandleDoubleAndSingleQuote`?

Reply:
HandleDoubleAndSingleQuote() will return the type of quote used as the first character.  So if the first character is not a quote it returns empty string.  In my case, I know that I need quotes so the difference is my method will always have it quoted.